packages/liveblocks-react/src/umbrella-store.ts

```diff
@@ -610,20 +612,32 @@ export class UmbrellaStore<M extends BaseMetadata> {
         );
       }
 
-      const result = await client.getInboxNotifications({ cursor });
+      try {
+        const result = await client.getInboxNotifications({ cursor });
 
-      this.updateThreadsAndNotifications(
-        result.threads as ThreadData<M>[], // TODO: Figure out how to remove this casting
-        result.inboxNotifications
-      );
+        this.updateThreadsAndNotifications(
+          result.threads as ThreadData<M>[], // TODO: Figure out how to remove this casting
+          result.inboxNotifications
+        );
 
-      // We initialize the `_lastRequestedNotificationsAt` date using the server timestamp after we've loaded the first page of inbox notifications.
-      if (this._notificationsLastRequestedAt === null) {
-        this._notificationsLastRequestedAt = result.requestedAt;
-      }
+        // We initialize the `_lastRequestedNotificationsAt` date using the server timestamp after we've loaded the first page of inbox notifications.
+        if (this._notificationsLastRequestedAt === null) {
+          this._notificationsLastRequestedAt = result.requestedAt;
+        }
+
+        const nextCursor = result.nextCursor;
+        return nextCursor;
+      } catch (err) {
+        // If the error is a 403 Forbidden error, we do not want to keep retrying
+        if (err instanceof NotificationsApiError && err.status === 403) {
+          throw new StopRetrying(
+            "403 Forbidden: Please check your permissions and try again."
+          );
```

Comment:
This error object will be thrown to the nearest `ErrorBoundary` on Suspense and will be used to populate the `error` object in non-suspense, so the message will say the following:

```
403 Forbidden: Please check your permissions and try again.
```

@nvie Please feel free to suggest alternative words that may be easier to understand for users.

Reply:
We could also instead do a simple:

```ts
throw new StopRetrying(err.message);
```

And, the message is normally:

```
Resource forbidden
```

Reply:
Hmm. I also realized this will expose `StopRetrying` error publicly. Does that mean we would need to also export `StopRetrying`? Maybe, time to think about a better error to expose 🤔

Reply:
I don't prefer catching and re-wrapping any existing errors like this, because it might become hard to trace where an error is coming from, what the real error/stack trace is, and where things are handled. Ideally, this function wouldn't have to be changed at all. Maybe a more solid approach is to start using the `throwError?: (err: any) => boolean` param from the `autoRetry` function, and somehow just define this logic in there.

For now, I'm okay hardcoding this `err instanceof NotificationsApiError && err.status === 403` check [where we set up the autoRetry call](https://github.com/liveblocks/liveblocks/blob/7e6c8cb485f9a827523c5d0402bc009b8430ea5b/packages/liveblocks-react/src/umbrella-store.ts#L478-L482), until we figure out a more flexible way of doing this, and do it similarly for `CommentsApiError` / `LiveblocksApiError`, etc. At least we maintain this ugly handling in a single place that way, and we don't have to spread it out across the code everywhere.

Reply:
> _"Please feel free to suggest alternative words that may be easier to understand for users."_

One of the things about our server error responses, is that they should absolutely contain the reason for the rejection in the HTTP response body, so ideally we would simply use that as the message. Ideally the various `*ApiError`s would expose those as `err.message` already, and we should simply reuse those.

Reply:
Good idea! I also do not like that `PaginatedResource` is now aware of `HttpError` but it is okay for now until we come up with better abstractions! Thanks for the pointer.

```ts
const initialFetcher = autoRetry(
  () => this._fetchPage(undefined),
  5,
  [5000, 5000, 10000, 15000],
  (err) => {
    // We do not want to retry if a 4xx HTTP error is received
    if (err instanceof HttpError && err.status >= 400 && err.status < 500) {
      return true;
    }
    return false;
  }
```

Reply:
I quite like this! Maybe we should make this PR part of `pagination-cleanups`, just to avoid any further merge conflicts there. We can just publish this fix a little bit later.

Reply:
Oh, nevermind, I just saw you already did that! Perfect! 💯